scripts/ml/2_finetune_gpl.py

```
@@ -373,62 +372,131 @@ def find_hard_negatives(
-
     all_ids = list(id_to_passage.keys())
+    triplets = []
 
-    for content_id, query_list in queries.items():
-        if content_id not in id_to_item:
-            continue
+    if model and embeddings:
+        # --- Per-query hard negative mining ---
+        # Each query gets its own hard negatives based on query-document similarity,
+        # rather than all queries for a document sharing the same negatives.
+        print(f"  Mining per-query hard negatives using base model...")
```

Ruff (0.15.5)

[error] f-string without any placeholders

Remove extraneous `f` prefix

(F541)